router/batchrouter/asyncdestinationmanager/salesforce-bulk/api_service.go

```diff
@@ -0,0 +1,279 @@
+	return apiErr
+}
+
+func (s *SalesforceAPIService) getCSVRecords(endpoint string) ([]map[string]string, *APIError) {
+	respBody, apiErr := s.makeRequest("GET", endpoint, bytes.NewReader([]byte{}), "")
+	if apiErr != nil {
+		return nil, apiErr
+	}
+
+	reader := csv.NewReader(bytes.NewReader(respBody))
+
+	headers, err := reader.Read()
+	if err == io.EOF {
+		return []map[string]string{}, nil
+	}
+	if err != nil {
+		return nil, &APIError{
+			StatusCode: 500,
+			Message:    fmt.Sprintf("reading CSV header: %v", err),
+			Category:   "ServerError",
+		}
+	}
+
+	var records []map[string]string
+	for {
+		row, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, &APIError{
+				StatusCode: 500,
+				Message:    fmt.Sprintf("reading CSV row: %v", err),
+				Category:   "ServerError",
+			}
+		}
+
+		record := make(map[string]string)
+		for i, value := range row {
+			if i < len(headers) {
+				record[headers[i]] = value
+			}
+		}
+		records = append(records, record)
+	}
+
+	return records, nil
+}
+
+func (s *SalesforceAPIService) makeRequest(
+	method, endpoint string,
+	body io.Reader,
+	contentType string,
+) ([]byte, *APIError) {
+	respBody, apiError := s.attemptRequest(method, endpoint, body, contentType)
+	if apiError == nil {
+		return respBody, nil
+	}
+
+	return nil, apiError
+}
+
+func (s *SalesforceAPIService) attemptRequest(
+	method, endpoint string,
+	body io.Reader,
+	contentType string,
+) ([]byte, *APIError) {
+	req, err := http.NewRequest(method, endpoint, body)
+	if err != nil {
+		return nil, &APIError{
+			StatusCode: 500,
+			Message:    fmt.Sprintf("creating HTTP request: %v", err),
+			Category:   "ServerError",
+		}
+	}
+
+	if contentType != "" {
+		req.Header.Set("Content-Type", contentType)
+	}
+
+	req = req.WithContext(cntx.CtxWithDestInfo(req.Context(), s.destinationInfo))
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return nil, &APIError{
+			StatusCode: 500,
+			Message:    fmt.Sprintf("making HTTP request: %v", err),
+			Category:   "ServerError",
+		}
+	}
+	defer resp.Body.Close()
+	respBody, err := io.ReadAll(resp.Body)
+	fmt.Println("response", string(respBody))
+	originalResponse := gjson.GetBytes(respBody, "originalResponse").String()
+	if originalResponse != "" {
+		respBody = []byte(originalResponse)
+	}
+	if err != nil {
+		return nil, &APIError{
+			StatusCode: 500,
+			Message:    fmt.Sprintf("reading response body: %v", err),
+			Category:   "ServerError",
+		}
+	}
+
+	if resp.StatusCode >= 400 {
+		return nil, &APIError{
+			StatusCode: resp.StatusCode,
+			Message:    string(respBody),
+			Category:   categorizeError(resp.StatusCode),
+		}
+	}
+
+	return respBody, nil
+}
+
+func categorizeError(statusCode int) string {
+	switch statusCode {
+	case 401:
+		return "RefreshToken"
+	case 429:
+		return "RateLimit"
+	case 400:
+		return "BadRequest"
+	case 500, 502, 503, 504:
+		return "ServerError"
+	default:
+		return "Unknown"
+	}
+}
```

Comment:
Let's use constants there for status codes.
```suggestion
func categorizeError(statusCode int) string {
	switch statusCode {
	case http.StatusUnauthorized:
		return "RefreshToken"
	case http.StatusTooManyRequests:
		return "RateLimit"
	case http.StatusBadRequest:
		return "BadRequest"
	case http.StatusInternalServerError, http.StatusBadGateway, http.StatusServiceUnavailable:
		return "ServerError"
	default:
		return "Unknown"
	}
}
```